crates/client/gateway/src/client/builder.rs

```diff
@@ -48,4 +76,170 @@ impl FeederClient {
+            } else {
+                None
+            }
+        };
+
+        // Wait for the pause duration if needed
+        if let Some(duration) = pause_duration {
+            tokio::time::sleep(duration).await;
+        }
+
+        Ok(f())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct RetryPolicy {
+    max_retries: usize,
+    backoff: Duration,
+    pause_until: Arc<RwLock<Option<Instant>>>,
+}
+
+impl RetryPolicy {
+    pub fn new(max_retries: usize, backoff: Duration, pause_until: Arc<RwLock<Option<Instant>>>) -> Self {
+        RetryPolicy { max_retries, backoff, pause_until }
+    }
+}
+
+impl Policy<Request<String>, Response<Incoming>, Box<dyn Error + Send + Sync>> for RetryPolicy {
+    type Future = Pin<Box<dyn Future<Output = Self> + Send>>;
+
+    fn retry(
+        &self,
+        _: &Request<String>,
+        result: Result<&Response<Incoming>, &Box<dyn Error + Send + Sync>>,
+    ) -> Option<Self::Future> {
+        let pause_until = self.pause_until.clone();
+
+        match result {
+            Ok(response) => {
+                // TOO_MANY_REQUESTS
+                if response.status().as_u16() == 429 {
+                    let retry_after = get_retry_after(response).unwrap_or(Duration::from_secs(10)); // Default 10 seconds
+
+                    let next_policy = self.clone();
+                    let fut = async move {
+                        let mut pause_lock = pause_until.write().await;
+                        *pause_lock = Some(Instant::now() + retry_after);
+
+                        // wait for the retry_after duration
+                        tokio::time::sleep(retry_after).await;
+
+                        next_policy
+                    };
+                    Some(Box::pin(fut))
+                } else {
+                    None
+                }
+            }
+            Err(_) if self.max_retries > 0 => {
+                // If the request failed, retry after backoff duration
+                let next_policy = RetryPolicy {
+                    max_retries: self.max_retries - 1,
+                    backoff: self.backoff,
+                    pause_until: self.pause_until.clone(),
+                };
+                let fut = tokio::time::sleep(self.backoff);
+                Some(Box::pin(async move {
+                    fut.await;
+                    next_policy
+                }))
+            }
+            _ => None, // No more retries
+        }
+    }
+
+    fn clone_request(&self, req: &Request<String>) -> Option<Request<String>> {
+        Some(req.clone())
+    }
+}
+
+fn get_retry_after(response: &Response<Incoming>) -> Option<Duration> {
+    if let Some(retry_after_header) = response.headers().get("Retry-After") {
+        if let Ok(retry_after_str) = retry_after_header.to_str() {
+            if let Ok(retry_seconds) = retry_after_str.parse::<u64>() {
+                return Some(Duration::from_secs(retry_seconds));
+            }
+        }
+    }
+    None
+}
```

Comment:
This seems counter-productive: what is the point of rate limiting if the request can pass a `Rety-After` header with a duration of 0s? Imo we should default all requests to a retry duration of 10s.